End disabling regions

paulfd opened the pull request:
Closes #365

paulfd committed:
Add a helper for playing voices

paulfd committed:
A region is disabled by end=-1

paulfd committed:
Disabled regions don't need to take up polyphony
The tests were a bit flaky and the did not seem justified by any behavior in e.g. ARIA
The new tests cover a similar situation which is validated in sforzando

jpcima commented on src/sfizz/Region.cpp:
```diff
@@ -92,7 +92,11 @@ bool sfz::Region::parseOpcode(const Opcode& rawOpcode)
+            disabled = true;
+        } else {
+            setValueFromOpcode(opcode, sampleEnd, Default::sampleEndRange);
+        }
```
We probably want to re-enable the region if sampleEnd is reset to a positive value after -1.
`<region>end=-1 end=10000`

Here I don't think the `disabled` boolean to be a great idea, I'd rather have `bool disabled() const` to check for `sampleEnd` where the member `sampleEnd` is changed to an optional.

Btw does `end` not also disable the voice if 0 ? I recall it to be the case.
Hence we will want to check `disabled` = `end` is set, and is <= 0.

paulfd replied:
Ok then I guess the best is to make it an int64, and create this `disabled()` function.

paulfd replied:
In the end I chose to keep the type. ARIA complains when sampleEnd==0 but still disables the region, so we can just do the same.

Otherwise I guess it'd be better to audit the uses of int64/uint32 in file and loop lengths. libsndfile uses int64 for the file length but the wav chunks are uint32.

paulfd committed:
Disable with sampleEnd == 0

paulfd committed:
Correct a compilation woe

jpcima commented on src/sfizz/Synth.cpp:
```diff
@@ -494,7 +494,7 @@ void sfz::Synth::finalizeSfzLoad()
             }
 
-            region->sampleEnd = std::min(region->sampleEnd, fileInformation->end);
+            region->sampleEnd = min(region->sampleEnd, fileInformation->end);
```
What is the reason for this edit here?

paulfd replied:
It's a leftover but I can revert it.

paulfd committed:
Add special cases for the rate transformation in envelopes
For increasing linear ramps, a 0s ramp is an increment of 1.0
For decaying exponential ramps, a 0s ramp is a multiplicator of 0.0f

paulfd committed:
Sustain tweaks:
- Separate the sustain value from the sustain threshold after which the envelope goes from decay to sustain
- If the sustain value is set to 0, the envelope is freerunning

paulfd commented:
The latest should match the behavior described in #365.

I amended the rate transformation methods for the cases where the `timeInSeconds` is exactly `0`, since by default the decaying envelope for example was actually quite long (around 1 000 samples).

I also believe this could fix #373 since the value for the sustain is taken as-is, and we don't add `virtuallyZero` to it anymore.
